### ISSUE TITLE
Karpenter tries to create 'null' instance if there is no capacity for a selected instance type in an AZ

### ISSUE THREAD

**Issue by Noksa** [aws/karpenter#1219]
## Version

**Karpenter**: v0.5.4

**Kubernetes**: v1.21.2

## Expected Behavior

Karpenter will try to create an instance type again and again until it is created
## Actual Behavior

Karpenter tries to create 'null' instance instead. See logs.
## Steps to Reproduce the Problem

Try to create a pod that requires an instance type that is not currently available in a desired AZ.
Pod should have/use volume in the desired AZ.
## Resource Specs and Logs

Logs:
```
2022-01-25T08:59:48.471Z	INFO	controller.provisioning	Batched 1 pods in 1.000893146s	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:48.656Z	DEBUG	controller.provisioning	Discovered subnets: [subnet-008e6f797aa1e359d (eu-central-1c) subnet-015d5bf7d03302f80 (eu-central-1a) subnet-05ac83c106785dcce (eu-central-1b)]	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:48.662Z	INFO	controller.provisioning	Computed packing of 1 node(s) for 1 pod(s) with instance type option(s) [c5a.4xlarge]	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:48.804Z	DEBUG	controller.provisioning	Discovered security groups: [sg-0eef900ebc61dd08b]	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:48.807Z	DEBUG	controller.provisioning	Discovered kubernetes version 1.21	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:48.877Z	DEBUG	controller.provisioning	Discovered ami ami-0764bbee991ae11a1 for query /aws/service/eks/optimized-ami/1.21/amazon-linux-2/recommended/image_id	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:48.877Z	DEBUG	controller.provisioning	Discovered caBundle, length 1025	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:48.914Z	DEBUG	controller.provisioning	Discovered launch template Karpenter-ad-eks-c-6892537386293631782	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:49.827Z	DEBUG	controller.provisioning	InsufficientInstanceCapacity for offering { instanceType: c5a.4xlarge, zone: eu-central-1b, capacityType: on-demand }, avoiding for 45s	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:49.827Z	ERROR	controller.provisioning	Could not launch node, launching instances, with fleet error(s), InsufficientInstanceCapacity: We currently do not have sufficient c5a.4xlarge capacity in the Availability Zone you requested (eu-central-1b). Our system will be working on provisioning additional capacity. You can currently get c5a.4xlarge capacity by not specifying an Availability Zone in your request or choosing eu-central-1a, eu-central-1c.	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:49.827Z	INFO	controller.provisioning	Waiting for unschedulable pods	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:50.828Z	INFO	controller.provisioning	Batched 1 pods in 1.000822137s	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:50.836Z	INFO	controller.provisioning	Computed packing of 1 node(s) for 1 pod(s) with instance type option(s) [c5a.4xlarge]	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:50.836Z	DEBUG	controller.provisioning	Discovered caBundle, length 1025	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:51.110Z	ERROR	controller.provisioning	Could not launch node, launching instances, with fleet error(s), InvalidFleetConfiguration: Your requested instance type (null) is not supported in your requested Availability Zone (eu-central-1c).; InvalidFleetConfiguration: Your requested instance type (null) is not supported in your requested Availability Zone (eu-central-1a).; InvalidFleetConfiguration: Your requested instance type (null) is not supported in your requested Availability Zone (eu-central-1b).	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:51.110Z	INFO	controller.provisioning	Waiting for unschedulable pods	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:55.828Z	INFO	controller.provisioning	Batched 1 pods in 1.000485974s	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:55.836Z	INFO	controller.provisioning	Computed packing of 1 node(s) for 1 pod(s) with instance type option(s) [c5a.4xlarge]	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:55.837Z	DEBUG	controller.provisioning	Discovered caBundle, length 1025	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:56.103Z	ERROR	controller.provisioning	Could not launch node, launching instances, with fleet error(s), InvalidFleetConfiguration: Your requested instance type (null) is not supported in your requested Availability Zone (eu-central-1b).; InvalidFleetConfiguration: Your requested instance type (null) is not supported in your requested Availability Zone (eu-central-1c).; InvalidFleetConfiguration: Your requested instance type (null) is not supported in your requested Availability Zone (eu-central-1a).	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T08:59:56.103Z	INFO	controller.provisioning	Waiting for unschedulable pods	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:02.104Z	INFO	controller.provisioning	Batched 1 pods in 1.000388744s	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:02.111Z	INFO	controller.provisioning	Computed packing of 1 node(s) for 1 pod(s) with instance type option(s) [c5a.4xlarge]	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:02.111Z	DEBUG	controller.provisioning	Discovered caBundle, length 1025	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:02.500Z	ERROR	controller.provisioning	Could not launch node, launching instances, with fleet error(s), InvalidFleetConfiguration: Your requested instance type (null) is not supported in your requested Availability Zone (eu-central-1b).; InvalidFleetConfiguration: Your requested instance type (null) is not supported in your requested Availability Zone (eu-central-1c).; InvalidFleetConfiguration: Your requested instance type (null) is not supported in your requested Availability Zone (eu-central-1a).	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:02.500Z	INFO	controller.provisioning	Waiting for unschedulable pods	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:08.503Z	INFO	controller.provisioning	Batched 1 pods in 1.000864735s	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:08.511Z	INFO	controller.provisioning	Computed packing of 1 node(s) for 1 pod(s) with instance type option(s) [c5a.4xlarge]	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:08.511Z	DEBUG	controller.provisioning	Discovered caBundle, length 1025	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:08.843Z	ERROR	controller.provisioning	Could not launch node, launching instances, with fleet error(s), InvalidFleetConfiguration: Your requested instance type (null) is not supported in your requested Availability Zone (eu-central-1c).; InvalidFleetConfiguration: Your requested instance type (null) is not supported in your requested Availability Zone (eu-central-1a).; InvalidFleetConfiguration: Your requested instance type (null) is not supported in your requested Availability Zone (eu-central-1b).	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:08.843Z	INFO	controller.provisioning	Waiting for unschedulable pods	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:14.844Z	INFO	controller.provisioning	Batched 1 pods in 1.000111361s	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:14.851Z	INFO	controller.provisioning	Computed packing of 1 node(s) for 1 pod(s) with instance type option(s) [c5a.4xlarge]	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:14.852Z	DEBUG	controller.provisioning	Discovered caBundle, length 1025	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:15.122Z	ERROR	controller.provisioning	Could not launch node, launching instances, with fleet error(s), InvalidFleetConfiguration: Your requested instance type (null) is not supported in your requested Availability Zone (eu-central-1c).; InvalidFleetConfiguration: Your requested instance type (null) is not supported in your requested Availability Zone (eu-central-1a).; InvalidFleetConfiguration: Your requested instance type (null) is not supported in your requested Availability Zone (eu-central-1b).	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:15.122Z	INFO	controller.provisioning	Waiting for unschedulable pods	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:21.123Z	INFO	controller.provisioning	Batched 1 pods in 1.000455927s	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:21.130Z	INFO	controller.provisioning	Computed packing of 1 node(s) for 1 pod(s) with instance type option(s) [c5a.4xlarge]	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:21.131Z	DEBUG	controller.provisioning	Discovered caBundle, length 1025	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
2022-01-25T09:00:21.412Z	ERROR	controller.provisioning	Could not launch node, launching instances, with fleet error(s), InvalidFleetConfiguration: Your requested instance type (null) is not supported in your requested Availability Zone (eu-central-1c).; InvalidFleetConfiguration: Your requested instance type (null) is not supported in your requested Availability Zone (eu-central-1a).; InvalidFleetConfiguration: Your requested instance type (null) is not supported in your requested Availability Zone (eu-central-1b).	{"commit": "7e79a67", "provisioner": "storage-db-smb-eu-central-1"}
```

**Comment by bwagner5** [comment]
Can you post the pod spec and provisioner spec? I'm a little confused on how it's getting `null`, hopefully the full configuration will allow me to reproduce this. :)

**Comment by Noksa** [comment]
Sure.

Provisioner:
```
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  annotations:
    meta.helm.sh/release-name: smb-eu-central-1
    meta.helm.sh/release-namespace: smb-eu-central-1
  creationTimestamp: "2022-01-08T05:39:28Z"
  generation: 8
  labels:
    app.kubernetes.io/managed-by: Helm
  name: storage-db-smb-eu-central-1
  resourceVersion: "183345574"
  uid: 3e707cf2-a6a0-4b06-b8ae-5d8675722cbc
spec:
  kubeletConfiguration: {}
  labels:
    only-db: "true"
  limits: {}
  provider:
    apiVersion: extensions.karpenter.sh/v1alpha1
    instanceProfile: ad-eks-c-NodeInstanceProfile
    kind: AWS
    securityGroupSelector:
      aws:eks:cluster-name: ad-eks-c
    subnetSelector:
      kubernetes.io/cluster/ad-eks-c: '*'
  requirements:
  - key: kubernetes.io/arch
    operator: In
    values:
    - amd64
  - key: node.kubernetes.io/instance-type
    operator: In
    values:
    - c5a.large
    - c5a.xlarge
    - c5a.2xlarge
    - c5a.4xlarge
  - key: karpenter.sh/capacity-type
    operator: In
    values:
    - on-demand
  taints:
  - effect: NoSchedule
    key: only-db
    value: "true"
  ttlSecondsAfterEmpty: 60
status:
  resources:
    cpu: "0"
    memory: "0"
```

Important parts from pod spec:
```
        resources:                                                                                                                                                                          │
          requests:                                                                                                                                                                         │
            cpu: "15"                                                                                                                                                                       │
            memory: 16Gi
...
      tolerations:                                                                                                                                                                          │
      - effect: NoSchedule                                                                                                                                                                  │
        key: only-db                                                                                                                                                                        │
        value: "true"
...
      nodeSelector:                                                                                                                                                                         │
        only-db: "true"
```

**Comment by ellistarn** [comment]
I think this was fixed in v0.5.5. Can you upgrade and try again?
https://github.com/aws/karpenter/commit/f8de85cdc68b59c8124cba9807bf9834a46488d1#diff-060f66d674e92171e12e4f7c94619d35ecda747c53cc792a3190d12bc0e58005
https://github.com/aws/karpenter/commit/54c6b37cd1b1c9e06410b4abca06baa751e85476#diff-060f66d674e92171e12e4f7c94619d35ecda747c53cc792a3190d12bc0e58005
